xpublish_edr/select.py

```diff
@@ -0,0 +1,125 @@
+    for query_param in query_param_keys:
+        if query_param in edr_query_params:
+            del query_params[query_param]
+
+    method: Optional[str] = "nearest"
+
+    for key, value in query_params.items():
+        split_value = value.split("/")
+        if len(split_value) == 1:
+            continue
+        elif len(split_value) == 2:
+            query_params[key] = slice(split_value[0], split_value[1])
+            method = None
+        else:
+            raise ValueError(f"Too many values for selecting {key}")
+
+    ds = ds.sel(query_params, method=method)
+    return ds
+
+
+def select_postition(ds: xr.Dataset, point: Point) -> xr.Dataset:
+    """
+    Return a dataset with the position nearest to the given coordinates
+    """
+    if _is_regular_xy_coords(ds):
+        return _select_position_regular_xy_grid(ds, point)
+    else:
+        # TODO: Handle 2D coordinates
+        raise NotImplementedError("Only 1D coordinates are supported")
+
+
+def select_area(ds: xr.Dataset, polygon: Polygon) -> xr.Dataset:
+    """
+    Return a dataset with the area within the given polygon
+    """
+    if _is_regular_xy_coords(ds):
+        return _select_area_regular_xy_grid(ds, polygon)
+    else:
+        # TODO: Handle 2D coordinates
+        raise NotImplementedError("Only 1D coordinates are supported")
+
+
+def _coord_is_regular(da: xr.DataArray) -> bool:
+    """
+    Check if the DataArray has a regular grid
+    """
+    return len(da.shape) == 1 and da.name in da.dims
+
+
+def _is_regular_xy_coords(ds: xr.Dataset) -> bool:
+    """
+    Check if the dataset has 2D coordinates
+    """
+    return _coord_is_regular(ds.cf["X"]) and _coord_is_regular(ds.cf["Y"])
+
+
+def _select_position_regular_xy_grid(ds: xr.Dataset, point: Point) -> xr.Dataset:
+    """
+    Return a dataset with the position nearest to the given coordinates
+    """
+    # Find the nearest X and Y coordinates to the point
+    return ds.cf.sel(X=point.x, Y=point.y, method="nearest")
+
+
+def _select_area_regular_xy_grid(ds: xr.Dataset, polygon: Polygon) -> xr.Dataset:
+    """
+    Return a dataset with the area within the given polygon
+    """
+    # For a regular grid, we can create a meshgrid of the X and Y coordinates to create a spatial mask
+    pts = np.meshgrid(ds.cf["X"], ds.cf["Y"])
+
+    # Create a mask of the points within the polygon
+    contains = np.vectorize(lambda p: polygon.contains(Point(p)), signature="(n)->()")
+    mask = contains(np.stack(pts, axis=-1))
+
+    # Find the x and y indices that have any points within the polygon
+    x_mask = mask.any(axis=0)
+    y_mask = mask.any(axis=1)
+
+    # Create a new DataArray with the mask matching the dataset dimensions
+    dims = ds.cf["Y"].dims + ds.cf["X"].dims
+    da_mask = xr.DataArray(mask, dims=dims)
+
+    # Apply the mask and subset out any data that does not fall within the polygon bounds
+    ds_subset = ds.where(da_mask).cf.isel(X=x_mask, Y=y_mask)
```

Comment:
Doesn;t this return a bunch of NaNs for say a pentagon?

IIUC what you need is
```
xinds, yinds = np.argwhere(mask)
return ds.cf.isel(X=xr.Variable(xinds, dims="points"), Y=xr.Variable(yinds, dims="points")
```

to trigger vectorized indexing

Reply:
Can xarray support a staggered array for this? The reason for the currenly impl is because i wasnt changing the coordinates, but maybe i should?

Reply:
Oh this is soooo cool. Thanks @dcherian !!!

Reply:
oops, I meant `np.nonzero` not `np.argwhere`